Avoid warning about not returning a value [blocks: #2548]

tautschnig opened the pull request:
Compilers differ in their ability to infer dead code, and the use of INVARIANT
etc. may result in end-of-function being reachable when they are configured
as no-ops. Make compilers happy, even when that may result in statements that
really shouldn't be reachable.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Remove unreachable return statement

All earlier branches end in a return statement.

tautschnig committed:
Remove unreachable return statement

Also clean up the earlier if(...) return true; return false;

kroening commented:
Won't that backfire, i.e., generate warnings about dead code?

tautschnig commented:
> Won't that backfire, i.e., generate warnings about dead code?

Not as far as I can tell, but maybe #2548 has other changes hiding such warnings, so we'll see when CI passes over here (where there really is just this one commit).

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0941f7b6382f696a8cc9bf4a77e2416c3c4dcc45).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91377050

peterschrammel reviewed (approved):
That's quite awkward. Hm... We could have a variant of these macros that take the return value to make this less ugly (but maybe also more confusing).

kroening commented:
I am wondering whether the goal here can be achieved by means of the right compiler annotation for "UNREACHABLE" for Visual Studio?
Or turn off the "missing return value" warning in UNREACHABLE?

smowton commented:
Some use of `__declspec(noreturn)` on a wrapper function ought to do the trick AFAICT

kroening commented:
@smowton these functions will return, just somewhere else, with a return statement. It needs to be an annotation that says "I promise we won't get here". It might be doable to do that with a throw on VS.

smowton commented:
? The important thing here is if we somehow hit an `UNREACHABLE` then control will not return to the callsite. That should stop VS from wanting a return statement.

kroening commented:
The relevant function already has a __declspec(noreturn). The problem appears to be that MSVC can't infer that
```
if(!(CONDITION))
   invariant_violated...(...);
```
doesn't have a successor for CONDITION = false.

kroening commented:
@tautschnig Special-casing the UNREACHABLE macro appears to work!

tautschnig commented:
@kroening I believe I have now implemented what you meant by special-casing, but please do confirm.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0aad92fbae70a88b193027d0f804e765b84a057d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95263703

tautschnig committed:
Fix wrong macro name and misplaced #endif

We previously never set noreturn for Visual Studio.

tautschnig committed:
Remove unreachable return

tautschnig committed:
Avoid warning about not returning a value

UNREACHABLE will throw an exception if such an instruction were actually
reached.

tautschnig commented:
Marking do-not-merge: I was likely mislead, the fix might be much simpler, investigating in #2310.

tautschnig commented:
@kroening @owen-jones-diffblue @smowton @kroening You might all want to take another look as I have completely reworked the commits in this PR. The proper fix was much simpler than the intermediate versions that I had, and to an extent boils down to fixing a single typo...

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c8054228a4caa12197191fe193d378c634e60cb0).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/95446748
